gtsam/hybrid/DCMixtureFactor.h

```diff
@@ -33,55 +33,92 @@ namespace gtsam {
  */
 template <class NonlinearFactorType>
 class DCMixtureFactor : public DCFactor {
+ public:
+  using Base = DCFactor;
+  using shared_ptr = boost::shared_ptr<DCMixtureFactor>;
+  using sharedFactor = boost::shared_ptr<NonlinearFactorType>;
+
+  /// typedef for DecisionTree which has Keys as node labels and
+  /// NonlinearFactorType as leaf nodes.
+  using Factors = DecisionTree<Key, sharedFactor>;
+
  private:
-  DiscreteKey dk_;
-  std::vector<NonlinearFactorType> factors_;
+  /// Decision tree of Gaussian factors indexed by discrete keys.
+  Factors factors_;
   bool normalized_;
 
  public:
-  using Base = DCFactor;
-
   DCMixtureFactor() = default;
 
-  DCMixtureFactor(const KeyVector& keys, const DiscreteKey& dk,
-                  const std::vector<NonlinearFactorType>& factors,
-                  bool normalized = false)
-      : dk_(dk), factors_(factors), normalized_(normalized) {
-    // Compiler doesn't like `keys_` in the initializer list.
-    keys_ = keys;
-
-    // Add `dk` to `dkeys` list.
-    discreteKeys_.push_back(dk);
-  }
+  /**
+   * @brief Construct from Decision tree.
+   *
+   * @param keys Vector of keys for continuous factors.
+   * @param discreteKeys Vector of discrete keys.
+   * @param factors Decision tree with of shared factors.
+   * @param normalized Flag indicating if the factor error is already
+   * normalized.
+   */
+  DCMixtureFactor(const KeyVector& keys, const DiscreteKeys& discreteKeys,
+                  const Factors& factors, bool normalized = false)
+      : Base(keys, discreteKeys), factors_(factors), normalized_(normalized) {}
 
-  /// Discrete key selecting mixture component
-  const DiscreteKey& discreteKey() const { return dk_; }
+  /**
+   * @brief Convenience constructor that generates the underlying factor
+   * decision tree for us.
+   *
+   * Here it is important that the vector of factors has the correct number of
+   * elements based on the number of discrete keys and the cardinality of the
+   * keys, so that the decision tree is constructed appropriately.
+   *
+   * @param keys Vector of keys for continuous factors.
+   * @param discreteKeys Vector of discrete keys.
+   * @param factors Vector of shared pointers to factors.
+   * @param normalized Flag indicating if the factor error is already
+   * normalized.
+   */
+  DCMixtureFactor(const KeyVector& keys, const DiscreteKeys& discreteKeys,
+                  const std::vector<sharedFactor>& factors,
+                  bool normalized = false)
+      : DCMixtureFactor(keys, discreteKeys,
+                        Factors(discreteKeys, factors)) {}
 
-  DCMixtureFactor& operator=(const DCMixtureFactor& rhs) {
-    Base::operator=(rhs);
-    this->dk_ = rhs.dk_;
-    this->factors_ = rhs.factors_;
+  /**
```

Comment:
Let's just kill. Caller should create shared pointers without copying

Reply:
I also found a bug in the `vector<sharedFactor>` version of the constructor. `normalized_` was not being set!